frontend/transformation-playground/src/context/PlaygroundContext.tsx

```diff
@@ -0,0 +1,217 @@
+import React, {
+  createContext,
+  useContext,
+  useReducer,
+  useEffect,
+  useMemo
+} from 'react';
+
+export interface InputDocument {
+  id: string;
+  name: string;
+  content: string;
+}
+
+export interface Transformation {
+  id: string;
+  name: string;
+  content: string;
+}
+
+export interface OutputDocument {
+  id: string;
+  name: string;
+  content: string;
+  sourceInputId: string;
+  transformationsApplied: string[]; // ids of all transformations that changed the document
+}
+
+export interface PlaygroundState {
+  inputDocuments: InputDocument[];
+  transformations: Transformation[];
+  outputDocuments: OutputDocument[];
+}
+
+const initialState: PlaygroundState = {
+  inputDocuments: [],
+  transformations: [],
+  outputDocuments: []
+};
+
+type ActionType =
+  | { type: 'ADD_INPUT_DOCUMENT'; payload: InputDocument }
+  | { type: 'UPDATE_INPUT_DOCUMENT'; payload: InputDocument }
+  | { type: 'REMOVE_INPUT_DOCUMENT'; payload: string }
+  | { type: 'ADD_TRANSFORMATION'; payload: Transformation }
+  | { type: 'UPDATE_TRANSFORMATION'; payload: Transformation }
+  | { type: 'REMOVE_TRANSFORMATION'; payload: string }
+  | {
+      type: 'REORDER_TRANSFORMATION';
+      payload: { fromIndex: number; toIndex: number };
+    }
+  | { type: 'ADD_OUTPUT_DOCUMENT'; payload: OutputDocument }
+  | { type: 'UPDATE_OUTPUT_DOCUMENT'; payload: OutputDocument }
+  | { type: 'REMOVE_OUTPUT_DOCUMENT'; payload: string }
+  | { type: 'CLEAR_OUTPUT_DOCUMENTS'; payload?: undefined }
+  | { type: 'SET_STATE'; payload: Partial<PlaygroundState> };
+
+const playgroundReducer = (
```

Comment:
Is this going to scale up over time? I'd worry about hard to discover bugs entering this function